### ISSUE TITLE
How do you loop a api call that depends on result given by the previous iteration

### ISSUE THREAD

**Issue by rubyphemmy** [redux-saga/redux-saga#1944]
### Description
I am very new to saga so this might be a noob question.
I have a function that is suppose to get all users in a given department with the following constraints:
- I don't know the number of users that the api will return
- The api only returns 100 results at a time.

Unfortunately, my function is not working as I planned, somehow, the `count` counter isn't incremented and both the `success` and `failure` actions are called at the first iteration then it stops. 

```js
const handleGetUsers = function*(action) {
  try {
    let response;
    let moreResults = true;
    let count = 0;

    while (moreResults) {
      const { dept, start, end } = action.payload;
      const res = yield call(CustomAxios.get, 'users', {
        params: {
          start,
          end,
          dept,
          limit: 100,
          skip: 100 * count,
        },
      });
      console.log('RES>>>>>>>>>>', res);
      Object.assign(response, res);
      if (res.data.data.length < 100) {
        moreResults = false;
      }

      count++;
    }
    console.log('RESPONSE>>>>>>', response);
    yield put(usersActions.getUsersSuccess(response.data.data || []));
  } catch (error) {
    yield put(showError({ content: 'Could not load users.' }));
    yield put(usersActions.getUsersFailure());
  }
};
```

**Comment by Andarist** [comment]
The code looks rather OK at first - you'd have to share your issue reproduced somewhere (e.g. codesandbox.io), without such it's hard to tell.

The only thing I've noticed is that probably this `Object.assign(response, res);` is buggy - because latest data can override what already got fetched. You need to introduce better merging logic there.

**Comment by armw4** [comment]
Yes it might be better to try `Object.assign({}, response, res);` to ensure you get [clean state per iteration](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign).

**Comment by Andarist** [comment]
It's not only that the `response` object gets mutated - depending on the `res` shape it's highly possible that it contains the same top-level keys each time so data from previous iteration might get lost (because it might be overridden). This is speculative though

**Comment by armw4** [comment]
Ah yes, that would be the opposite of accumulation (e.g. arrays are additive to you would not have to take such special care) `[myObject, ...spreadTheOtherArray]`

**Comment by rubyphemmy** [comment]
@Andarist You are right the `object.assign` portion of the code was malfunctioning. I replaced it with `..spread` and it worked fine.

**Comment by Andarist** [comment]
@rubyphemmy can this issue be closed then?
